### PR TITLE
feat: Kickoff Transformation implementationtransformation code base

### PR DESCRIPTION
**Pull request by HaoXuAI**
# What this PR does / why we need it:

Created a Transformation interface. it still works with the current pandas_transformation, python_transformation etc.

The next step is refactor the BatchMaterializationEngine to make it works for both Materialization and Transformation.

# Which issue(s) this PR fixes:

https://github.com/feast-dev/feast/issues/4584
https://github.com/feast-dev/feast/issues/4277#issuecomment-2427772928
https://github.com/feast-dev/feast/issues/4696
# Misc

### CODE REVIEW

**Comment by HaoXuAI** [comment]
Reference comment: https://github.com/feast-dev/feast/pull/5130#issuecomment-2742126583

**Comment by HaoXuAI** [comment]
One unstable unit test seems to be related to HF:
`FAILED sdk/python/tests/unit/test_on_demand_python_transformation.py::TestOnDemandTransformationsWithWrites::test_docling_transform - huggingface_hub.errors.LocalEntryNotFoundError: An error happened while trying to locate the file on the Hub and we cannot find the requested files in the local cache. Please check your connection and try again or make sure your Internet connection is on.`

**Comment by franciscojavierarceo** [comment]
Got it, I can make a patch for that. 👍

**Comment by HaoXuAI** [comment]
> Got it, I can make a patch for that. 👍

No rush, take your time. This pr is ready for review now :)

File: sdk/python/feast/on_demand_feature_view.py [hunk 1]
```diff
@@ -678,9 +687,6 @@ def _construct_random_input(
     ) -> dict[str, Union[list[Any], Any]]:
         rand_dict_value: dict[ValueType, Union[list[Any], Any]] = {
             ValueType.BYTES: [str.encode("hello world")],
-            ValueType.PDF_BYTES: [
```
Comment (franciscojavierarceo) [inline sdk/python/feast/on_demand_feature_view.py:681]: This is what's making the unit tests fail. 

This is used in `infer_features` to validate the function schema / types actually.

File: sdk/python/feast/on_demand_feature_view.py [hunk 2]
```diff
@@ -786,38 +792,22 @@ def mainify(obj) -> None:
...
-                raise TypeError(
-                    f"return signature for {user_function} is {return_annotation} but should be ibis.expr.types.relations.Table"
-                )
-            transformation = SubstraitTransformation.from_ibis(user_function, sources)
 
         on_demand_feature_view_obj = OnDemandFeatureView(
             name=name if name is not None else user_function.__name__,
             sources=sources,
             schema=schema,
-            feature_transformation=transformation,
```
Comment (franciscojavierarceo) [inline sdk/python/feast/on_demand_feature_view.py:813]: shouldn't we keep this to be backwards compatible?

File: sdk/python/feast/transformation/base.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    UserDefinedFunctionV2 as UserDefinedFunctionProto,
+)
+from feast.transformation.factory import (
+    TRANSFORMATION_CLASS_FOR_TYPE,
+    get_transformation_class_from_type,
+)
+from feast.transformation.mode import TransformationMode
+
+
+class Transformation(ABC):
```
Comment (franciscojavierarceo) [inline sdk/python/feast/transformation/base.py:20]: nit: I"d add a docstring here. ChatGPT should be a nice friend.

File: sdk/python/feast/transformation/base.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+def transformation(
+    mode: Union[TransformationMode, str],
+    name: Optional[str] = None,
+    tags: Optional[Dict[str, str]] = None,
+    description: Optional[str] = "",
+    owner: Optional[str] = "",
+):
+    def mainify(obj):
+        # Needed to allow dill to properly serialize the udf. Otherwise, clients will need to have a file with the same
```
Comment (franciscojavierarceo) [inline sdk/python/feast/transformation/base.py:99]: Hmm, we encountered some serialization issues with dill in the past. @rostifar do you remember what they were?

File: sdk/python/feast/transformation/python_transformation.py [hunk 5]
```diff
@@ -58,19 +109,11 @@ def infer_features(
                         f"Failed to infer type for feature '{feature_name}' with value "
                         + f"'{feature_value}' since no items were returned by the UDF."
                     )
+                inferred_type = type(feature_value[0])
                 inferred_value = feature_value[0]
-                if singleton and isinstance(inferred_value, list):
```
Comment (franciscojavierarceo) [inline sdk/python/feast/transformation/python_transformation.py:62]: I just added this FYI as it's an edge case I didn't consider.

**Review by franciscojavierarceo — COMMENTED** [review]
this look great! some small nits and some notes about fixing stuff I recently added which will fix the unit tests. otherwise looks awesome. 👏🚀🤠

File: sdk/python/feast/transformation/python_transformation.py (hunk 5 above)
Comment (HaoXuAI) [reply sdk/python/feast/transformation/python_transformation.py:62]: strange I didn't modify this. not sure why it sneak in

File: sdk/python/feast/on_demand_feature_view.py (hunk 2 above)
Comment (HaoXuAI) [reply sdk/python/feast/on_demand_feature_view.py:813]: Oh I guess that's merge conflict and I delete it accidently. let me add it back

Comment (HaoXuAI) [reply sdk/python/feast/on_demand_feature_view.py:813]: It is backwards compatible. I put the feature_transformation extraction logic inside the OnDemandFeatureView initialization, since the decorator doesn't pass in the feature_transformation param.
So user can do:
```
@demand_feature_view(...)
def udf()...
```
or
```
odfv = OnDemandFeatureView(feature_transformation=Transformation(...))`
```

File: sdk/python/feast/on_demand_feature_view.py (hunk 1 above)
Comment (HaoXuAI) [reply sdk/python/feast/on_demand_feature_view.py:681]: Ah I think I accidentally delete it when resolve merge conflict. let me add it back

**Comment by HaoXuAI** [comment]
> this look great! some small nits and some notes about fixing stuff I recently added which will fix the unit tests. otherwise looks awesome. 👏🚀🤠

good catch. thanks!

File: sdk/python/feast/transformation/base.py (hunk 3 above)
Comment (HaoXuAI) [reply sdk/python/feast/transformation/base.py:20]: That's good point

**Comment by HaoXuAI** [comment]
@franciscojavierarceo finally passed all tests!

File: sdk/python/feast/transformation/base.py (hunk 4 above)
Comment (HaoXuAI) [reply sdk/python/feast/transformation/base.py:123]: have dill issue as well. Created a new issue for it:
https://github.com/feast-dev/feast/issues/5182

**Comment by franciscojavierarceo** [comment]
> @franciscojavierarceo finally passed all tests!

image

**Review by franciscojavierarceo — APPROVED** [review]
Awesome work @HaoXuAI 🚀🚀🚀
